R/fetch.R

```diff
@@ -162,24 +162,36 @@ git_clone <- function(url, path = NULL, branch = NULL, password = askpass, ssh_k
 #' @param rebase if TRUE we try to rebase instead of merge local changes. This
 #' is not possible in case of conflicts (you will get an error).
 #' @param ... arguments passed to [git_fetch]
-git_pull <- function(rebase = FALSE, ..., repo = '.'){
+git_pull <- function(remote = NULL, rebase = FALSE, ..., repo = '.'){
   repo <- git_open(repo)
   info <- git_info(repo)
-  upstream <- info$upstream
+  branch <- info$shorthand
+  if (branch == "HEAD")
+    stop("Repository is currently in a detached head state")
+
+  upstream <- if(length(remote) && nchar(remote)){
+    paste0(remote, '/', branch)
```

Comment:
I feel like this is not consistent with what `git pull` does when the remote is specified but no refspec is given.

https://git-scm.com/docs/git-pull#_default_behaviour

> The rule to determine which remote branch to merge after fetching is a bit involved, in order not to break backward compatibility.
>
> If explicit refspecs were given on the command line of git pull, they are all merged.
> 
> When no refspec was given on the command line, then git pull uses the refspec from the configuration or $GIT_DIR/remotes/<origin>. In such cases, the following rules apply:
> 
> If branch.<name>.merge configuration for the current branch <name> exists, that is the name of the branch at the remote site that is merged.
>
> If the refspec is a globbing one, nothing is merged.
>
> Otherwise the remote branch of the first refspec is merged.

I pay close attention to this because, in usethis's PR flow, it's common for the name of the local branch != the remote branch.

Here's the remote and branch config for an external PR branch that supports pull and push:

```
[remote "avalcarcel9"]
	url = git@github.com:avalcarcel9/usethis.git
	fetch = +refs/heads/*:refs/remotes/avalcarcel9/*
	created-by = usethis::pr_fetch
[branch "avalcarcel9-add_use_author"]
	remote = avalcarcel9
	merge = refs/heads/add_use_author
	created-by = usethis::pr_fetch
	pr-url = https://github.com/r-lib/usethis/pull/833
```

Even with the addition of `remote`, I don't think usethis can use `gert::git_pull()`. I think I still have to use an internal DIY version to deal with local branch name != remote branch name. I think I'm basically creating my own way to give a refspec.

https://github.com/r-lib/usethis/blob/58e68b01957fc23bea42d0c47bbb01a0b68b2a8d/R/utils-git.R#L176-L203

Reply:
I understand not wanting to implement all of that from scratch. Not sure how much one could get by doing less and letting git config be consulted.

But it's possible that leaving `git_pull()` "as is" is the right call, but then I think its potential to behave differently from `git pull` needs to be clearly documented, with the recommendation to do fetch + merge in more complicated situations.

Reply:
Would it help if we change it so that you specify an upstream branch instead of just the remote? I.e. you do `git_pull("jeroen/mybranch")` instead of `git_pull("jeroen")` ?

Reply:
That *would* certainly handle my use case. Of course, I have the luxury of solving a specific problem and being the only caller.

But yeah it does seem like the ability to specify remote **and branch** would be nice. Remote alone doesn't really help me.

Could you get this "for free", i.e. does libgit2 honour git config?

> > If branch.\<name\>.merge configuration for the current branch \<name\> exists, that is the name of the branch at the remote site that is merged.

Or would you have to do it yourself?

Reply:
In my example above, the branch has its `remote` and the remote branch fully specified in the config. And that's intentional, i.e. so that `git pull`and the RStudio "pull" button do the right thing.